[fix] replace embedded HTML by data_src

return42 opened the pull request:
## What does this PR do?

[mod] result_templates/**videos.html** replace embedded HTML by data_src
[mod] result_templates/**default.html** replace embedded HTML by data_src

Embedded HTML breaks SearXNG architecture.  To modularize, HTML is generated in
the templates (oscar & simple) and result parameter 'embedded' is replaced by
'data_src', an URL for embedded content (<iframe>).

[mod] hostname_replace: replace hostnames in result's data_src param

## How to test this PR locally?

The (iframe & audio) templates of the following engines has been touched (oscar & simple theme)

- searx/engines/dailymotion.py
- searx/engines/invidious.py
- searx/engines/peertube.py --> https://github.com/searxng/searxng/issues/881
- searx/engines/sepiasearch.py
- searx/engines/vimeo.py
- searx/engines/youtube_api.py
- searx/engines/bandcamp.py
- searx/engines/deezer.py
- searx/engines/freesound.py
- searx/engines/mixcloud.py
- searx/engines/soundcloud.py

To test you need to redirect embedded videos (e.g.) from youtube to a invidios
instance.  Search for videos using engine `!youtube lebowski`.  The result URLs
and the embedded videos should link to the invidios instance.

Here is an example of such a `hostname_replace` configuration::

```yaml

    hostname_replace:

      # youtube --> Invidious

      '(.*\.)?youtube-nocookie\.com': 'invidio.xamh.de'
      '(.*\.)?youtube\.com$': 'invidio.xamh.de'
      '(.*\.)?invidious\.snopyta\.org$': 'invidio.xamh.de'
      '(.*\.)?vid\.puffyan\.us': 'invidio.xamh.de'
      '(.*\.)?invidious\.kavin\.rocks$': 'invidio.xamh.de'
      '(.*\.)?inv\.riverside\.rocks$': 'invidio.xamh.de'
```

## Related issues

Closes: https://github.com/searxng/searxng/issues/873

return42 committed:
[mod] result_templates/videos.html: replace embedded HTML by data_src

Embedded HTML breaks SearXNG architecture.  To modularize, HTML is generated in
the templates (oscar & simple) and result parameter 'embedded' is replaced by
'data_src', an URL for embedded content (<iframe>).

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[mod] hostname_replace: replace hostnames in result's data_src param

To test you need to redirect embeded videos (e.g.) from youtube to a invidios
instance.  Search for videos using engine `!youtube lebowski`.  The result URLs
and the embeded videos should link to the invidios instance.

Here is an example of such a `hostname_replace` configuration::

    hostname_replace:

      # youtube --> Invidious

      '(.*\.)?youtube-nocookie\.com': 'invidio.xamh.de'
      '(.*\.)?youtube\.com$': 'invidio.xamh.de'
      '(.*\.)?invidious\.snopyta\.org$': 'invidio.xamh.de'
      '(.*\.)?vid\.puffyan\.us': 'invidio.xamh.de'
      '(.*\.)?invidious\.kavin\.rocks$': 'invidio.xamh.de'
      '(.*\.)?inv\.riverside\.rocks$': 'invidio.xamh.de'

Closes: https://github.com/searxng/searxng/issues/873
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[mod] result_templates/default.html replace embedded HTML by data_src audio_src

Embedded HTML breaks SearXNG architecture.  To modularize, HTML is generated in
the templates (oscar & simple) and result parameter 'embedded' is replaced by
'data_src' (and 'audio_src'), an URL for embedded content (<iframe>).

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

dalf commented on searx/engines/vimeo.py:
```diff
@@ -64,7 +57,7 @@ def response(resp):
                 'template': 'videos.html',
                 'publishedDate': publishedDate,
-                'embedded': embedded,
+                'data_src': "https://player.vimeo.com/video/{videoid}" + videoid,
```
`{videoid}` must be removed

dalf commented on searx/plugins/hostname_replace.py:
```diff
@@ -28,5 +28,15 @@ def on_result(request, search, result):
                 return False
             result[parsed] = result[parsed]._replace(netloc=pattern.sub(replacement, result[parsed].netloc))
             result['url'] = urlunparse(result[parsed])
+        if result.get('data_src', False):
```
Nit: `if result.get('data_src'):` should be enough since `if None` is evaluate as `if False`.

return42 replied:
The entire commit needed to be reimplemented, read commit message from 55e2daf8b28

dalf commented on searx/plugins/hostname_replace.py:
```diff
@@ -28,5 +28,15 @@ def on_result(request, search, result):
+            if pattern.search(parsed_data_src.netloc):
+                parsed_data_src = parsed_data_src._replace(netloc=pattern.sub(replacement, parsed_data_src.netloc))
+                result['data_src'] = urlunparse(parsed_data_src)
+        if result.get('audio_src', False):
```
Nit: `if result.get('audio_src'):`

dalf commented on searx/static/themes/simple/src/less/style.less:
```diff
@@ -296,10 +296,26 @@ article[data-vim-selected].category-social {
 }
 
+.audio-control audio {
+  width: 100%;
```
See comment below

dalf commented on searx/templates/oscar/result_templates/default.html:
```diff
@@ -3,13 +3,13 @@
-   {{- result.embedded|safe -}}
+{% if result.data_src -%}
+<div id="result-media-{{ index }}" class="embedded-content invisible">
+  <iframe data-src="{{result.data_src}}" frameborder="0" allowfullscreen></iframe>
```
I would add `scrolling="no"` to avoid scroll bars.

The attribute is [deprecated](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe#attr-scrolling), but `overflow: hidden` is not enough for the iframe.

dalf commented on searx/templates/oscar/result_templates/videos.html:
```diff
@@ -3,15 +3,15 @@
-   {{- result.embedded|safe -}}
+{% if result.data_src -%}
+<div id="result-video-{{ index }}" class="embedded-video collapse">
+  <iframe data-src="{{result.data_src}}" width="540" height="304" frameborder="0" allowfullscreen></iframe>
```
I would add `scrolling="no"` to avoid scroll bars.

return42 replied:
fixed in a rebase https://github.com/searxng/searxng/pull/882/files#diff-149bcd9955091769a827c269387a306bba70bcf9e8e619cf4c6840611b846d51R12

dalf commented on searx/templates/simple/result_templates/default.html:
```diff
@@ -15,9 +15,14 @@
-   {{- result.embedded|safe -}}
+{% if result.data_src -%}
+<div id="result-media-{{ index }}" class="embedded-content invisible">
+  <iframe data-src="{{result.data_src}}" frameborder="0" allowfullscreen></iframe>
```
I would add `scrolling="no"` to avoid scroll bars.

return42 replied:
Set `scrolling="no"`  in oscar is OK for me, but in simple theme I think we should control the layout throughout in CSS.

dalf commented on searx/static/themes/simple/src/less/style.less:
```diff
@@ -296,10 +296,26 @@ article[data-vim-selected].category-social {
 
+.result-videos .embedded-video iframe {
+  width: 100%;
+  aspect-ratio: 16 / 9;
```
The property is not supported by one year old browser: https://caniuse.com/?search=aspect-ratio

A possible fallback:

```less
@supports not (aspect-ratio: 1 / 1) {
  .result-videos .embedded-video iframe {
    height: calc(@results-width * 9 / 16); // the right value
  }

  @media screen and (max-width: @tablet) {
    .result-videos .embedded-video iframe {
      height: 20rem; // A guess, the video should be displayed full screen anyway...
    }
  }

  @media screen and (max-width: @phone) {
    .result-videos .embedded-video iframe {
      height: auto; // A should, the video should be displayed full screen anyway...
    }
  }
}
```

For reference: https://css-tricks.com/aspect-ratio-boxes/

return42 replied:
> The property is not supported by one year old browser ..

I think ...

```yaml
@supports not (aspect-ratio: 1 / 1) {
  .result-videos .embedded-video iframe {
    height: calc(@results-width * 9 / 16); // the right value
  }
}
```

is an good addition to support older browsers (thanks for this suggestion).

But I don't want to implement different screen solutions, at least not in this PR.  The intention of this PR is not to get perfect layout we hadn't before, the intention is mostly to modularize HTML in templates.  We can improve the layout afterwards and discuss better layouts, BTW the aspect ratio of videos is not always 16:9 and on small devices it might be better when the "show video" button shows the video at a full screen.

return42 replied:
FYI I added edc31d84

dalf reviewed (changes requested):
See my comments

return42 committed:
[fix] hostname_replace.py: don't stop replace URL in fields

This is a rewrite of the hostname_replace.py that:

- don't stop to replace URL in fields ('data_src', 'audio_src') if there isn't a
  'parsed_url',
- adds a comment about keep or remove a result from the result list
- adds a loop over ['data_src', 'audio_src'] instead of doubling code lines

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[oscar] add scrolling="no" to iframes in HTML templates

Suggested-by: @dalf https://github.com/searxng/searxng/pull/882#discussion_r805180787
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[fix] simple theme: support browsers which do not have aspect-ratio

Suggested-by: @dalf https://github.com/searxng/searxng/pull/882#discussion_r805187303
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[mod] templates: rename field for <iframe> URL to iframe_src

Rename result field data_src to iframe_src

Suggested-by: @dalf https://github.com/searxng/searxng/pull/882#issuecomment-1037997402
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

dalf commented on searx/static/themes/simple/src/less/style.less:
```diff
@@ -324,6 +324,26 @@ article[data-vim-selected].category-social {
+  // 200px, somtimes 250px.
+  height: 250px;
+}
+
```
Without additional CSS classes on iframe:
```less
a.result.soundcloud iframe {
  height: 120px;
}

a.result.deezer iframe {
  height: 94px;
}

a.result.mixcloud iframe {
  height: 250px;
}
```

assertion: if the `hostname_replace` plugin replace some URL, the widget will have the same height.

If the property is based on the hostname, then each time the `hostname_replace` plugin configuration change, the theme has to be recompiled to add the new host.

---

There is a bug that can be fixed in another PR:
https://github.com/searxng/searxng/blob/a4942213a1dc6404616e239e0d2fb55a6e19a499/searx/templates/simple/macros.html#L22

the spaces of the engine name must be replaced by underscore.

return42 replied:
> assertion: if the hostname_replace plugin replace some URL, the widget will have the same height.

I think both concept do not really scale with players from a hostname-replace. These media players from different hostname might have a different layout (compare youtube an invidious video players).

I added a commit 2f806ea9 to have a custom-layer for CSS / I know the price is one request more for the additional `css/custom.css` but I do not know a better solution that fits into any concept.  IMO the layout of players is related to their URL (netloc) and by this we need to be able to layout by URL (netloc).

dalf commented on searx/webapp.py:
```diff
@@ -496,6 +496,7 @@ def render(template_name: str, override_theme: str = None, **kwargs):
 
     # helpers to create links to other pages
     kwargs['url_for'] = url_for_theme  # override url_for function in templates
+    kwargs['urlparse'] = urllib.parse.urlparse
```
If some reason the  CSS attribute selector are not a good choice (?), we can reintroduce the `extract_domain` filter:
https://github.com/searx/searx/blob/0248777f959ba9ee26fc8f1c367ea4ef2cb5d7cc/searx/webapp.py#L285-L288

return42 replied:
I didn't aware of CSS attribute selector / thanks for pointing out.

I removed this commit,

dalf commented on searx/static/themes/simple/src/less/custom.less:
```diff
@@ -1,4 +1,4 @@
-iframe.soundcloud\.com {
+iframe.w\.soundcloud\.com {
```
The host names and other CSS class names are overlapping. While the chance there is an actual overlap is small, it would be better to avoid to take the risk.
 
It is possible to use CSS attribute selector: https://developer.mozilla.org/en-US/docs/Web/CSS/Attribute_selectors

```less
iframe[src^="https://w.soundcloud.com"] {
  height: 120px;
}

iframe[src^="https://www.deezer.com"] {
  // The real size is 92px, but 94px are needed to avoid an inner scrollbar of
  // the embedded HTML.
  height: 94px;
}

iframe[src^="https://www.mixcloud.com"] {
  // the embedded player from mixcloud has some quirks: initial there is an
  // issue with an image URL that is blocked since it is an a Cross-Origin
  // request. The alternative text (<img alt='Mixcloud Logo'> then cause an
  // scrollbar in the inner of the iframe we can't avoid.  Another quirk comes
  // when pressing the play button, somtimes the shown player has an height of
  // 200px, somtimes 250px.
  height: 250px;
}
```

So there is no need
* to add `kwargs['urlparse'] = urllib.parse.urlparse`
* to call `urlparse` from `video.html` and `default.html` (iframe and audio element).

( changes done in commit 1d72ada65dafafeca99cbca45a549e384897f18f )

return42 replied:
Thanks, implemented in 8acf540f

dalf commented on searx/plugins/hostname_replace.py:
```diff
@@ -17,16 +17,30 @@
+                result[parsed] = result[parsed]._replace(netloc=pattern.sub(replacement, result[parsed].netloc))
+                result['url'] = urlunparse(result[parsed])
+
+        if not replacement:
```
Currently it is okay.
However, some engines return YouTube videos (like qwant, google video, bing video ?)
Currently the `iframe_src` attribute is not set for these engines.
If in the future, this is attribute is set , this condition is going to prevent the hostname replacement in the `iframe_src` attribute.

return42 replied:
I rewrote this commit, see 795e8af6

dalf commented on searx/plugins/hostname_replace.py:
```diff
@@ -17,16 +17,30 @@
+            if result.get(url_field):
+                url_src = urlparse(result[url_field])
+                if pattern.search(url_src.netloc):
+                    url_src = url_src._replace(netloc=pattern.sub(replacement, url_src.netloc))
```
```python
if pattern.search(url_src.netloc):
  if remplacement:
    del result[url_field]
  else:
    url_src = url_src._replace(netloc=pattern.sub(replacement, url_src.netloc))
    result[url_field] = urlunparse(url_src)
```

otherwise the code is going to crash on line 43 if `replacement` is `False` (which is allowed in the configuration).

return42 replied:
I rewrote this commit, see 795e8af6

dalf commented on searx/static/themes/simple/gruntfile.js:
```diff
@@ -138,6 +138,12 @@ module.exports = function (grunt) {
+            src: ['src/less/custom.less'],
+            dest: 'css/custom.css',
+            nonull: true,
+            filter: file_exists,
```
I can't see the benefit of an additional CSS file: `custom.css` is built file like the other built files => `custom.less` won't be customize because it requires to build the themes, so `git pull` is not going to work.

So, despite what I've said before, and after a second thought, I suggest to make it simple:
* rename custom.less to embedded.less
* move the CSS rules about embedded media from style.less to embedded.less ( `.embedded-content iframe {`, `.audio-control audio {`, etc...)
* include embedded.less in the normal build process (if it is not already done).
* leave the customization for later.

So everything about embedded media is in one place.

return42 replied:
> include embedded.less in the normal build process (if it is not already done).

A ``@import "embedded.less"`` in `searx/static/themes/simple/src/less/style.less` is needed  .. see 8acf540f

dalf replied:
Nit: I would have move the CSS rules about embedded media from style.less to embedded.less ( .embedded-content iframe {, .audio-control audio {, etc...)

dalf reviewed (changes requested):
After few days rereading the code, I end up to think there is no need to for a custom.less file (for now).

The code inside this file can simplified, see the comments.

After that, I think the PR is ready.

return42 committed:
[mod] simple theme: set height of embedded content (iframe)

This commit sets appropriate height of the (embedded) player from:

- soundcloud
- mixcloud
- deezer

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[build] /static

dalf reviewed (approved):
Thank you for this contribution.
It fixes an long standing issue.

---

In addition to the test described, I have check this configuration:
```yaml
hostname_replace:
  '(.*\.)?youtube-nocookie\.com': false
```
👍 it works
